v3/src/components/graph/adornments/adornments.tsx

```diff
@@ -0,0 +1,124 @@
+                           ? layout.plotHeight / yCatLabels.length
+                           : layout.plotHeight
+    // The instanceKey is used by the adornment model to uniquely identify an instance of the adornment in its
+    // map of instances. If there is only one instance of the adornment, the instanceKey is an empty string.
+    const instanceKey = xCatLabels.length === 0 && yCatLabels.length === 0
+        ? ''
+        : xCatLabels.length > 0
+          ? yCatLabels.length > 0
+            ? `{x: ${xCatLabels[index % xCatLabels.length]}, y: ${yCatLabels[Math.floor(index / xCatLabels.length)]}}`
+            : `{x: ${xCatLabels[index]}}`
+          : `{y: ${yCatLabels[index]}}`
+    // The adornmentKey is a unique value used for React's key prop. We can't use the instanceKey because that
+    // value may be duplicated if there are multiple adornment types active on the graph.
+    const adornmentKey = `${adornment.id}${instanceKey ? `-${instanceKey}` : ''}`
+    // TODO: Handle other types of adornments
+    if (adornment.type === "Movable Line") {
+      return (
+        <MovableLine
+          key={adornmentKey}
+          lineKey={instanceKey}
+          model={adornment as IMovableLineModel}
+          plotHeight={subPlotHeight}
+          plotIndex={index}
+          plotWidth={subPlotWidth}
+          xAxis={graphModel.getAxis('bottom') as INumericAxisModel}
+          yAxis={graphModel.getAxis('left') as INumericAxisModel}
+        />
+      )
+    }
+    return null
+  }
+
+  // When a graph contains multiple sub-plots, each adornment needs to be rendered once
+  // per sub-plot. We create a CSS grid in which to place the adornments, with each cell
+  // of the grid corresponding to one sub-plot.
+  const { xCategories, yCategories } = setCategories(),
+    colCount = xCategories.length,
+    rowCount = yCategories.length,
+    { left, top, width, height } = layout.computedBounds.plot,
+    cellsRequired = (colCount > 0 ? colCount : 1) * (rowCount > 0 ? rowCount : 1),
+    gridTemplateColumns = `repeat(${colCount}, 1fr)`,
+    gridTemplateRows = `repeat(${rowCount}, 1fr)`,
+    gridStyle = {
+      gridTemplateColumns,
+      gridTemplateRows,
+      height: `${height}px`,
+      left,
+      top,
+      width: `${width}px`,
+    }
+  let cellsRendered = 0
+
+  const renderAdornments = () => {
+    const adornmentNodes = []
+    while (cellsRendered < cellsRequired) {
+      adornmentNodes.push(
+        <div key={`graph-adornments-cell-${cellsRendered}`} className="graph-adornments-grid__cell">
+          {
+            adornments.map((adornment: IAdornmentModelUnion) => {
+              return setAdornmentNode(adornment, cellsRendered, xCategories, yCategories)
+            })
+          }
+        </div>
+      )
+      cellsRendered++
+    }
+    return adornmentNodes
+  }
+
+  return (
+    <div className={kGraphAdornmentsClass} style={gridStyle}>
+      {renderAdornments()}
```

Comment:
This is a matter of stylistic preference, but I would eliminate the `renderAdornments` function and just put its contents inline here. The `Adornments` functional component renders its contents by looping over the adornments, having it defer to another function seems unnecessary.

Reply:
@kswenson Is the change I made for this what you had in mind? I believe the while loop has to be outside of the return statement, so I couldn't put the full contents of `renderAdornments` here. I'm not sure if that's exactly what you meant, though.

Reply:
@kswenson Is [the change I made for this](https://github.com/concord-consortium/codap/blob/23bee14cd3834b0b9df1cbcc1844dcb3ff38dda7/v3/src/components/graph/adornments/adornments.tsx#L63-L87) what you had in mind? I believe the while loop has to happen outside/before the return, so I couldn't put the full contents of `renderAdornments` there exactly, but I'm not sure that's exactly what you meant.

Reply:
I'm not sure what I had in mind would have worked exactly, but this seems like an improvement.